### PR TITLE
SONARIAC-279 Rule S6387: Azure role assignments that grant access to all resources of a subscription are security-sensitive

### PR DESCRIPTION
**Pull request by nils-werner-sonarsource**

### CODE REVIEW

File: iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  private RoleScopeHelper() {
+    // helper class
+  }
+
+  public static Predicate<String> matchPredicate(String pattern) {
+    return s -> Pattern.matches(pattern, s);
+  }
+
+  public static boolean isSensitiveScope(ExpressionTree scope, Predicate<String> referenceScopePredicate, Predicate<String> plainScopePredicate) {
+    if (scope instanceof AttributeAccessTree) {
```
Comment (andrea-guarino-sonarsource) [inline iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java:51]: _Minor_: what about using `scope.is(ATTRIBUTE_ACCESS)` instead of `instanceof` ?

File: iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      .filter(TemplateInterpolationTree.class::isInstance)
+      .map(interpolation -> ((TemplateInterpolationTree) interpolation).expression())
+      .filter(AttributeAccessTree.class::isInstance)
+      .map(AttributeAccessTree.class::cast)
+      .anyMatch(interpolation -> containsSensitiveScope(interpolation, referenceScopePredicate));
+  }
+
+  public static String referenceToString(AttributeAccessTree reference) throws IllegalArgumentException {
+    StringBuilder sb = new StringBuilder();
+    if (reference.object() instanceof AttributeAccessTree) {
```
Comment (andrea-guarino-sonarsource) [inline iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java:75]: same here, we could probably replace `instanceof` with `is`

File: iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    } else if (reference.object() instanceof VariableExprTree) {
+      sb.append(((VariableExprTree) reference.object()).value());
+      sb.append('.');
+    }
+    sb.append(reference.attribute().value());
+    return sb.toString();
+  }
+
+  private static boolean isLimitedToResourceGroup(TemplateExpressionTree scope) {
+    return (scope).parts().stream()
```
Comment (andrea-guarino-sonarsource) [inline iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java:87]: unnecessary parens around `scope`?

File: iac-extensions/terraform/src/test/resources/checks/Azure/SubscriptionOwnerCapabilitiesCheck/compliant.tf [hunk 4]
```diff
@@ -33,6 +33,6 @@ resource "azurerm_role_definition" "role-with-limited-assignable-scopes-alternat
   }
 
   assignable_scopes = [
-    "/subscriptions/${var.subscription_id}/resourceGroups/${var.environment_name}" # Limited to a resource group
+    "/subscriptions/${data.azurerm_subscription.primary.id}/resourceGroups/${var.environment_name}" # Limited to a resource group
```
Comment (andrea-guarino-sonarsource) [inline iac-extensions/terraform/src/test/resources/checks/Azure/SubscriptionOwnerCapabilitiesCheck/compliant.tf:36]: I don't fully understand why did you change this existing test

**Review by andrea-guarino-sonarsource — APPROVED** [review]
LGTM, I left a few minor comments

File: iac-extensions/terraform/src/test/resources/checks/Azure/SubscriptionOwnerCapabilitiesCheck/compliant.tf (hunk 4 above)
Comment (nils-werner-sonarsource) [reply iac-extensions/terraform/src/test/resources/checks/Azure/SubscriptionOwnerCapabilitiesCheck/compliant.tf:36]: Sorry for the lack of explanation. The content of this check is to check if a relation is created to a resource and contains the keyword `resourceGroups` in it's path. The focus here is on the keyword `resourceGroups`. I had to change the test, because we don't resolve variables and therefore I entered the resolved relation of the resource to be checked. If the keyword would be missing, then we would have a false negative, or the keyword check would not work, because the relation is not provided as a basic requirement.

Comment (andrea-guarino-sonarsource) [reply iac-extensions/terraform/src/test/resources/checks/Azure/SubscriptionOwnerCapabilitiesCheck/compliant.tf:36]: Ok I see, thanks for clarifying!
